numpy/linalg/linalg.py

```diff
@@ -1300,19 +1300,21 @@ def svd(a, full_matrices=True, compute_uv=True):
     """
     Singular Value Decomposition.
 
-    Factors the matrix `a` as ``u * np.diag(s) * v``, where `u` and `v`
-    are unitary and `s` is a 1-d array of `a`'s singular values.
+    Factors the matrix `a` as ``u @ np.diag(s) @ vh``. When `a` is a 2D array,
```

Comment:
Hmm, `diag` is a nasty function. Could write this as ` (u * s) @ vh` (non- stacked), or `(u * s[..., None, :]) @ vh` for the stacked case. I don't know if that is any clearer, though.

Reply:
Which is to say, that those two forms are greatly preferred for actual code, but might not be the best from a pedagogical standpoint.

Reply:
Might put those alternatives in the `Notes` section.

Reply:
`diag` fails in confusing ways for the stacked case, which is an argument against using it.

These are correct too, right?
* `u @ (s[...,None] * vh)`
* `u @ (s[...,None]*np.eye(s.shape[-1]) @ vh`

Reply:
Those examples also look correct. I prefer the original as they are more likely to be cache friendly, i.e., contiguous access.

Reply:
Although the right hand versions look cleaner, so might be better for newbies.

Reply:
Could also introduce (in another PR) a `np.linalg.diag` with signature `(...,N) -> (...,N,N)` that behaves correctly on stacks of vectors, and doesn't have the weird overloading that `np.diag` has

Reply:
What about first explaining the non-stacked case, both with `diag` and the more efficient approach?

Further down, the stacked case can be discussed. Because that is never really going to be trivial in terms of python code, I would directly go for `u @ (s[...,None] * vh)` and skip the `eye` method. (I also prefer the right-hand version because it is a little shorter.)

Reply:
I think keeping a special case for 2d using `diag` is probably the right thing to do.

> as they are more likely to be cache friendly, i.e., contiguous access.

Are you sure about that? Are the `vh` and `u` matrices fortran-order?

Reply:
They all seem to be C contiguous.

Reply:
The comments on the contiguity and caching are not completely obvious to me. Both `s[..., None, :]` and `s[..., None]` are C contiguous, which seems fine, so I guess the problem is in the broadcasting of the element-wise multiplication.

Say we have a matrix `a` with shape `(5, 4, 3)`:

- Case `(u * s[..., None, :])`. I guess `u` is traversed contiguously, so 4 passes through `s` are needed when computing this product.

- Case `(s[..., None] * vh)`. I guess `vh` is traversed contiguously, so only 1 pass through `s` is needed when computing this product.

Isn't the second case better for caching? (I'm not sure how Numpy loops through the arrays internally.)

Are  such performance issues relevant compared to the cost of SVD?

Reply:
Yeah, it isn't clear how the actions are actually ordered, but one way would be to skip down the columns of `vh` multiplying by the elements of `s`, the skipping breaking the cache, while in the first case the elements would all be accessed sequentially in memory, which would be cache friendly. But as you say, it could be that the rows of `vh` might be multiplied by a single element, which would also be cache friendly. In practice, it looks like the second is a bit better
```
In [1]: u = array([eye(3)]*1000000)

In [2]: s = array([[1,2,3]]*1000000)

In [3]: %timeit s[...,None]*u
10 loops, best of 3: 30.2 ms per loop

In [4]: %timeit u*s[...,None,:]
10 loops, best of 3: 32.8 ms per loop
```
Talking about cache is probably out of scope for a docstring, so I would go with the example that looks best. Both are more efficient than converting the singular values to matrices ...

Reply:
Thanks for explaining. I'll further edit the docs in line with the discussions here.